File: golang-devildyw/middle/L199二叉树的右视图.go
package main

func main() {}

func rightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	result := []int{}
	// 层序遍历 那我就层序遍历呗
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		length := len(queue)
		for length > 0 {
			//出队
			first := queue[0]
			queue = queue[1:]
			if length == 1 {
				result = append(result, first.Val)
			}
			if first.Left != nil {
				queue = append(queue, first.Left)
			}
			if first.Right != nil {
				queue = append(queue, first.Right)
			}
			length--
		}
	}
	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L199二叉树的右视图.java
package top.dyw.leetcode.Hot100;

public class L199二叉树的右视图 {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Deque<TreeNode> deque = new LinkedList<>();
        deque.offer(root);
        while(!deque.isEmpty()) {

            int n = deque.size();

            for (int i = 0; i<n; i++) {
                TreeNode node = deque.poll();
                if (node.left != null) {
                    deque.offer(node.left);
                }
                if (node.right != null) {
                    deque.offer(node.right);
                }

                if (i==n-1) {
                    result.add(node.val);
                }
            }
        }
        return result;
    }
}
